"Multiple diagnostics in flight at once!" assert, and runaway memory usage in release builds, on C++20 testcase with PCH

version :  llvmorg-17.0.3 

Recently, when using clangd in vscode remote, once a certain file is encountered, the memory will explode, eventually causing the entire machine to go out of memory (clangd consumes more than 64G of memory). Below is the stack (using the release version, compiled with -g)

```
#0  0x00007f4d2d1a8233 in _int_malloc () from /lib64/libc.so.6
#1  0x00007f4d2d1ab10c in malloc () from /lib64/libc.so.6
#2  0x0000000000a2b2d8 in llvm::SmallVectorBase<unsigned int>::grow_pod(void*, unsigned long, unsigned long) ()
#3  0x0000000000f35070 in llvm::SmallVectorTemplateCommon<std::pair<clang::DiagnosticsEngine::ArgumentKind, long>, void>::grow_pod (TSize=16, MinSize=<optimized out>, this=0x7f4d25fe3710) at /llvmorg-17.0.3/llvm/include/llvm/ADT/SmallVector.h:140
#4  llvm::SmallVectorTemplateBase<std::pair<clang::DiagnosticsEngine::ArgumentKind, long>, true>::grow (MinSize=<optimized out>, this=0x7f4d25fe3710) at /llvmorg-17.0.3/llvm/include/llvm/ADT/SmallVector.h:529
#5  llvm::SmallVectorTemplateCommon<std::pair<clang::DiagnosticsEngine::ArgumentKind, long>, void>::reserveForParamAndGetAddressImpl<llvm::SmallVectorTemplateBase<std::pair<clang::DiagnosticsEngine::ArgumentKind, long>, true> > (N=1, Elt=<synthetic pointer>..., This=0x7f4d25fe3710)
    at /llvmorg-17.0.3/llvm/include/llvm/ADT/SmallVector.h:246
#6  llvm::SmallVectorTemplateBase<std::pair<clang::DiagnosticsEngine::ArgumentKind, long>, true>::reserveForParamAndGetAddress (N=1, Elt=<synthetic pointer>..., this=0x7f4d25fe3710) at /llvmorg-17.0.3/llvm/include/llvm/ADT/SmallVector.h:541
#7  llvm::SmallVectorTemplateBase<std::pair<clang::DiagnosticsEngine::ArgumentKind, long>, true>::push_back (Elt=..., this=0x7f4d25fe3710) at /llvmorg-17.0.3/llvm/include/llvm/ADT/SmallVector.h:566
#8  clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x6a4068 <(anonymous namespace)::StaticDiagInfoDescriptions+249544> "'", DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1142
#9  0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#10 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x6a3d23 <(anonymous namespace)::StaticDiagInfoDescriptions+248707> ")' clause%select{s|}1", DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#11 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#12 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x6a3cf0 <(anonymous namespace)::StaticDiagInfoDescriptions+248656> " cannot be mixed with '%0(%select{sink:vec|source}1)' clause%select{s|}1", DiagEnd=<optimized out>, OutStr=...)
    at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#13 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#14 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x6a3cdb <(anonymous namespace)::StaticDiagInfoDescriptions+248635> ")' clause%select{|s}1 cannot be mixed with '%0(%select{sink:vec|source}1)' clause%select{s|}1", DiagEnd=<optimized out>, OutStr=...)
    at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#15 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#16 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x6a2e66 <(anonymous namespace)::StaticDiagInfoDescriptions+244934> "", DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#17 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#18 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x6a138a <(anonymous namespace)::StaticDiagInfoDescriptions+238058> "", DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#19 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#20 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x6a08f0 <(anonymous namespace)::StaticDiagInfoDescriptions+235344> " type", DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#21 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#22 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x69fb28 <(anonymous namespace)::StaticDiagInfoDescriptions+231816> ", not %0", DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#23 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#24 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x69f1b4 <(anonymous namespace)::StaticDiagInfoDescriptions+229396> " because assigning method's 2nd parameter of type %0 is not an Objective-C pointer type", DiagEnd=<optimized out>, OutStr=...)
    at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#25 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#26 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x69ea67 <(anonymous namespace)::StaticDiagInfoDescriptions+227527> " element must have Objective-C object return type instead of %0", DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#27 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#28 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x69e6bf <(anonymous namespace)::StaticDiagInfoDescriptions+226591> " %3", DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#29 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#30 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x69e660 <(anonymous namespace)::StaticDiagInfoDescriptions+226496> " declaration conflicts with previous %select{|direct }2declaration of %select{method|property}1 %3", DiagEnd=<optimized out>, OutStr=...)
    at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#31 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#32 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x69e3eb <(anonymous namespace)::StaticDiagInfoDescriptions+225867> "", DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#33 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#34 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x69c9ab <(anonymous namespace)::StaticDiagInfoDescriptions+219147> "", DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#35 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#36 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x69bde6 <(anonymous namespace)::StaticDiagInfoDescriptions+216134> "", DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#37 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#38 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x69b4bd <(anonymous namespace)::StaticDiagInfoDescriptions+213789> " follows declaration in %select{the global module|module %4}3", DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#39 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#40 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x69b239 <(anonymous namespace)::StaticDiagInfoDescriptions+213145> " method %0 not found ; did you mean %2?", DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#41 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#42 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x69ad1b <(anonymous namespace)::StaticDiagInfoDescriptions+211835> ", but function has %select{non-const lvalue|rvalue}2 ref-qualifier", DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#43 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#44 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x69aa5c <(anonymous namespace)::StaticDiagInfoDescriptions+211132> " %2?", DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#45 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#46 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x69a6ec <(anonymous namespace)::StaticDiagInfoDescriptions+210252> " integers and enum types, but was given %2%select{| whose underlying type is %4}3", DiagEnd=<optimized out>, OutStr=...)
    at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
   .......
   .......
   ....... something like FormatDiagnostic
   .......
   .......
   .......
   .......
#224 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930, DiagStr=0x67f46c <(anonymous namespace)::StaticDiagInfoDescriptions+99020> " found %select{%select{no super class|super class with type %5}4|instance variable '%4' access control is %select{|@private|@protected|@public|@package}5}3",
    DiagEnd=<optimized out>, OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#225 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#226 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930,
    DiagStr=0x67f30c <(anonymous namespace)::StaticDiagInfoDescriptions+98668> " has different definitions in different modules; first difference is this %select{||||static assert|field|method|type alias|typedef|data member|friend declaration|function template|method|instance var"..., DiagEnd=<optimized out>,
    OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#227 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#228 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930,
    DiagStr=0x67f1e8 <(anonymous namespace)::StaticDiagInfoDescriptions+98376> " found %select{end of class|public access specifier|private access specifier|protected access specifier|static assert|field|method|type alias|typedef|data member|friend declaration|function template|m"..., DiagEnd=<optimized out>,
    OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#229 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#230 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930,
    DiagStr=0x67f064 <(anonymous namespace)::StaticDiagInfoDescriptions+97988> " found %select{%select{method %5|constructor|destructor}4 that has %6 parameter%s6|%select{method %5|constructor|destructor}4 with %ordinal6 parameter of type %7%select{| decayed from %9}8|%select{met"..., DiagEnd=<optimized out>,
    OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#231 0x0000000000f35375 in HandleSelectModifier (OutStr=..., ArgumentLen=<optimized out>, Argument=<optimized out>, ValNo=<optimized out>, DInfo=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:628
#232 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930,
    DiagStr=0x67eeea <(anonymous namespace)::StaticDiagInfoDescriptions+97610> " first difference is %select{return type is %4|%ordinal4 parameter with name %5|%ordinal4 parameter with type %5%select{| decayed from %7}6|%ordinal4 parameter with%select{out|}5 a default argument|%o"..., DiagEnd=<optimized out>,
    OutStr=...) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:1001
#233 0x0000000001de8f76 in clang::clangd::fillNonLocationData (DiagLevel=clang::DiagnosticsEngine::Error, Info=..., D=...) at /llvmorg-17.0.3/clang-tools-extra/clangd/Diagnostics.cpp:664
#234 0x0000000001dea974 in operator() (__closure=__closure@entry=0x7f4d25ff66e0, D=...) at /llvmorg-17.0.3/clang-tools-extra/clangd/Diagnostics.cpp:720
#235 0x0000000001df8222 in clang::clangd::StoreDiags::HandleDiagnostic (this=0x7f4d25ff9be0, DiagLevel=<optimized out>, Info=...) at /opt/gcc-11.1.0/include/c++/11.1.0/optional:440
#236 0x0000000000f3a572 in clang::DiagnosticIDs::EmitDiag (DiagLevel=clang::DiagnosticIDs::Error, Diag=..., this=0x7f4d1c01e1f0) at /llvmorg-17.0.3/clang/lib/Basic/DiagnosticIDs.cpp:823
#237 clang::DiagnosticIDs::ProcessDiag (this=this@entry=0x7f4d1c01e1f0, Diag=...) at /llvmorg-17.0.3/clang/lib/Basic/DiagnosticIDs.cpp:815
#238 0x0000000000f33f1b in clang::DiagnosticsEngine::ProcessDiag (this=0x7f4d1c001990) at /llvmorg-17.0.3/clang/include/clang/Basic/Diagnostic.h:1037
#239 clang::DiagnosticsEngine::EmitCurrentDiagnostic (this=0x7f4d1c001990, Force=<optimized out>) at /llvmorg-17.0.3/clang/lib/Basic/Diagnostic.cpp:545
#240 0x0000000000b61590 in clang::DiagnosticBuilder::~DiagnosticBuilder() ()
#241 0x0000000000e0f1f4 in clang::ODRDiagsEmitter::diagnoseMismatch(clang::FunctionDecl const*, clang::FunctionDecl const*) const ()
#242 0x0000000002b5b877 in clang::ASTReader::diagnoseOdrViolations() [clone .localalias] ()
#243 0x0000000002b5e761 in clang::ASTReader::FinishedDeserializing() [clone .part.0] ()
#244 0x0000000002bbac65 in clang::ASTReader::ReadDeclRecord(unsigned int) ()
#245 0x0000000002b4d6c9 in clang::ASTReader::GetDecl(unsigned int) [clone .localalias] ()
#246 0x0000000002b665c3 in clang::ASTReader::completeVisibleDeclsMap(clang::DeclContext const*) [clone .part.0] ()
#247 0x00000000018c2e1e in (anonymous namespace)::LookupVisibleHelper::lookupInDeclContext(clang::DeclContext*, clang::LookupResult&, bool, bool) ()
#248 0x00000000018c5282 in clang::Sema::LookupVisibleDecls(clang::DeclContext*, clang::Sema::LookupNameKind, clang::VisibleDeclConsumer&, bool, bool, bool) [clone .localalias] ()
#249 0x00000000018cf4f4 in clang::Sema::makeTypoCorrectionConsumer(clang::DeclarationNameInfo const&, clang::Sema::LookupNameKind, clang::Scope*, clang::CXXScopeSpec*, clang::CorrectionCandidateCallback&, clang::DeclContext*, bool, clang::ObjCObjectPointerType const*, bool) [clone .localalias] ()
#250 0x00000000018d0c1d in clang::Sema::CorrectTypo(clang::DeclarationNameInfo const&, clang::Sema::LookupNameKind, clang::Scope*, clang::CXXScopeSpec*, clang::CorrectionCandidateCallback&, clang::Sema::CorrectTypoKind, clang::DeclContext*, bool, clang::ObjCObjectPointerType const*, bool) ()
#251 0x0000000001536b02 in clang::Sema::DiagnoseUnknownTypeName(clang::IdentifierInfo*&, clang::SourceLocation, clang::Scope*, clang::CXXScopeSpec*, clang::OpaquePtr<clang::QualType>&, bool) ()
#252 0x0000000002a1a27d in clang::Parser::ParseImplicitInt(clang::DeclSpec&, clang::CXXScopeSpec*, clang::Parser::ParsedTemplateInfo const&, clang::AccessSpecifier, clang::Parser::DeclSpecContext, clang::ParsedAttributes&) [clone .localalias] ()
#253 0x0000000002a1e49c in clang::Parser::ParseDeclarationSpecifiers(clang::DeclSpec&, clang::Parser::ParsedTemplateInfo const&, clang::AccessSpecifier, clang::Parser::DeclSpecContext, clang::Parser::LateParsedAttrList*, clang::ImplicitTypenameContext) [clone .localalias] ()
#254 0x0000000002af8dad in clang::Parser::ParseDeclOrFunctionDefInternal(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec&, clang::AccessSpecifier) [clone .localalias] ()
#255 0x0000000002af9e9f in clang::Parser::ParseDeclarationOrFunctionDefinition(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*, clang::AccessSpecifier) [clone .part.0] ()
#256 0x0000000002b00676 in clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*) [clone .localalias] ()
#257 0x0000000002b018da in clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) [clone .localalias] ()
#258 0x00000000029fe3ca in clang::ParseAST(clang::Sema&, bool, bool) [clone .localalias] ()
#259 0x000000000116f5f9 in clang::FrontendAction::Execute() ()
#260 0x0000000001ee75fd in clang::clangd::ParsedAST::build (Filename=..., Inputs=..., CI=..., CompilerInvocationDiags=..., Preamble=...) at /llvmorg-17.0.3/clang-tools-extra/clangd/ParsedAST.cpp:669
#261 0x0000000001f5d462 in clang::clangd::(anonymous namespace)::ASTWorker::generateDiagnostics (this=0x3946950, Invocation=..., Inputs=..., CIDiags=...) at /opt/gcc-11.1.0/include/c++/11.1.0/bits/basic_string.h:912
#262 0x0000000001f5dd83 in operator() (__closure=0x7f4cf80022d0) at /opt/gcc-11.1.0/include/c++/11.1.0/bits/unique_ptr.h:172
#263 0x0000000001f56965 in llvm::unique_function<void ()>::operator()() (this=0x3946bb0) at /llvmorg-17.0.3/llvm/include/llvm/ADT/FunctionExtras.h:382
#264 llvm::function_ref<void ()>::callback_fn<llvm::unique_function<void ()> >(long) (callable=60058544) at /llvmorg-17.0.3/llvm/include/llvm/ADT/STLFunctionalExtras.h:45
#265 llvm::function_ref<void ()>::operator()() const (this=<synthetic pointer>) at /llvmorg-17.0.3/llvm/include/llvm/ADT/STLFunctionalExtras.h:68
#266 clang::clangd::(anonymous namespace)::ASTWorker::runTask(llvm::StringRef, llvm::function_ref<void()>) (this=this@entry=0x3946950, Name=..., Task=...) at /llvmorg-17.0.3/clang-tools-extra/clangd/TUScheduler.cpp:1324
#267 0x0000000001f58439 in clang::clangd::(anonymous namespace)::ASTWorker::run (this=0x3946950) at /llvmorg-17.0.3/clang-tools-extra/clangd/TUScheduler.cpp:1458
#268 0x0000000002106f25 in llvm::unique_function<void ()>::operator()() (this=0x392abc0) at /llvmorg-17.0.3/llvm/include/llvm/ADT/FunctionExtras.h:382
#269 operator() (__closure=0x392aba0) at /llvmorg-17.0.3/clang-tools-extra/clangd/support/Threading.cpp:101
#270 operator()<clang::clangd::AsyncTaskRunner::runAsync(const llvm::Twine&, llvm::unique_function<void()>)::<lambda()>&> (F=..., __closure=<optimized out>) at /llvmorg-17.0.3/llvm/include/llvm/Support/thread.h:43
#271 std::__invoke_impl<void, llvm::thread::GenericThreadProxy<std::tuple<clang::clangd::AsyncTaskRunner::runAsync(const llvm::Twine&, llvm::unique_function<void()>)::<lambda()> > >(void*)::<lambda(auto:4&&, auto:5&& ...)>, clang::clangd::AsyncTaskRunner::runAsync(const llvm::Twine&, llvm::unique_function<void()>)::<lambda()>&> (__f=...) at /opt/gcc-11.1.0/include/c++/11.1.0/bits/invoke.h:61
#272 std::__invoke<llvm::thread::GenericThreadProxy<std::tuple<clang::clangd::AsyncTaskRunner::runAsync(const llvm::Twine&, llvm::unique_function<void()>)::<lambda()> > >(void*)::<lambda(auto:4&&, auto:5&& ...)>, clang::clangd::AsyncTaskRunner::runAsync(const llvm::Twine&, llvm::unique_function<void()>)::<lambda()>&> (__fn=...) at /opt/gcc-11.1.0/include/c++/11.1.0/bits/invoke.h:96
#273 std::__apply_impl<llvm::thread::GenericThreadProxy<std::tuple<clang::clangd::AsyncTaskRunner::runAsync(const llvm::Twine&, llvm::unique_function<void()>)::<lambda()> > >(void*)::<lambda(auto:4&&, auto:5&& ...)>, std::tuple<clang::clangd::AsyncTaskRunner::runAsync(const llvm::Twine&, llvm::unique_function<void()>)::<lambda()> >&, 0> (__t=..., __f=...) at /opt/gcc-11.1.0/include/c++/11.1.0/tuple:1806
#274 std::apply<llvm::thread::GenericThreadProxy<std::tuple<clang::clangd::AsyncTaskRunner::runAsync(const llvm::Twine&, llvm::unique_function<void()>)::<lambda()> > >(void*)::<lambda(auto:4&&, auto:5&& ...)>, std::tuple<clang::clangd::AsyncTaskRunner::runAsync(const llvm::Twine&, llvm::unique_function<void()>)::<lambda()> >&> (__t=..., __f=...) at /opt/gcc-11.1.0/include/c++/11.1.0/tuple:1817
#275 llvm::thread::GenericThreadProxy<std::tuple<clang::clangd::AsyncTaskRunner::runAsync(const llvm::Twine&, llvm::unique_function<void()>)::<lambda()> > > (Ptr=0x392aba0) at /llvmorg-17.0.3/llvm/include/llvm/Support/thread.h:41
#276 llvm::thread::ThreadProxy<std::tuple<clang::clangd::AsyncTaskRunner::runAsync(const llvm::Twine&, llvm::unique_function<void()>)::<lambda()> > >(void *) (Ptr=0x392aba0) at /llvmorg-17.0.3/llvm/include/llvm/Support/thread.h:55
#277 0x00007f4d2e258e25 in start_thread () from /lib64/libpthread.so.0
#278 0x00007f4d2d22335d in clone () from /lib64/libc.so.6

```

the  input to `clang::Diagnostic::FormatDiagnostic` look like this:

```
 f 232
#232 clang::Diagnostic::FormatDiagnostic (this=0x7f4d25ff6930,
    DiagStr=0x67eeea <(anonymous namespace)::StaticDiagInfoDescriptions+97610> " first difference is %select{return type is %4|%ordinal4 parameter with name %5|%ordinal4 parameter with type %5%select{| decayed from %7}6|%ordinal4 parameter with%select{out|}5 a default argument|%o"..., DiagEnd=<optimized out>,
    OutStr=...) at /source/install/llvm/llvm.git/clang/lib/Basic/Diagnostic.cpp:1001
1001            HandleSelectModifier(*this, Val, Argument, ArgumentLen, OutStr);
(gdb) p DiagStr
$5 = 0x67eeea <(anonymous namespace)::StaticDiagInfoDescriptions+97610> " first difference is %select{return type is %4|%ordinal4 parameter with name %5|%ordinal4 parameter with type %5%select{| decayed from %7}6|%ordinal4 parameter with%select{out|}5 a default argument|%o"...
(gdb) printf "%s\n", DiagStr
 first difference is %select{return type is %4|%ordinal4 parameter with name %5|%ordinal4 parameter with type %5%select{| decayed from %7}6|%ordinal4 parameter with%select{out|}5 a default argument|%ordinal4 parameter with a default argument|function body}3
(gdb)

```

and the `OutStr` contains the hugh memory

```
(gdb) p OutStr
$24 = (llvm::SmallVectorImpl<char> &) @0x7f4d25ff6500: {
  <llvm::SmallVectorTemplateBase<char, true>> = {
    <llvm::SmallVectorTemplateCommon<char, void>> = {
      <llvm::SmallVectorBase<unsigned long>> = {
        BeginX = 0x7f48d43fa010,
        Size = 4511505832,
        Capacity = 8724152319
      }, <No data fields>},
    members of llvm::SmallVectorTemplateBase<char, true>:
    static TakesParamByValue = true
  }, <No data fields>}
(gdb)

```

the content of `OutStr` 



```
(gdb) x /20xs (0x7f48d43fa010)
0x7f48d43fa010: "'std::midpoint' has different definitions in different modules; 1 first difference is '_Tp'"
0x7f48d43fa06c: " has different definitions in different modules; first difference is 1 found '_Tp'"
0x7f48d43fa0bf: " has different definitions in different modules; first difference is 1 found '_Tp'"
0x7f48d43fa112: " 1 has different definitions in different modules; first difference is this '_Tp'"
0x7f48d43fa164: " has different definitions in different modules; first difference is 1 found '_Tp'"
0x7f48d43fa1b7: " has different definitions in different modules; first difference is 1 found '_Tp'"
0x7f48d43fa20a: " has different definitions in different modules; first difference is 1 found '_Tp'"
0x7f48d43fa25d: " has different definitions in different modules; first difference is 1 found '_Tp'"
0x7f48d43fa2b0: " has different definitions in different modules; first difference is 1 found '_Tp'"
0x7f48d43fa303: " has different definitions in different modules; first difference is 1 found '_Tp'"
0x7f48d43fa356: " has different definitions in different modules; first difference is 1 found '_Tp'"
0x7f48d43fa3a9: " has different definitions in different modules; first difference is 1 found '_Tp'"
0x7f48d43fa3fc: "template parameter lists have a different number of parameters ( vs 2)"
0x7f48d43fa443: "template parameter has different kinds in different translation units"
0x7f48d43fa489: "field  declared with incompatible types in different translation units (2 vs. 1)"
0x7f48d43fa4da: "external function  declared with incompatible types in different translation units (2 vs. 1)"
0x7f48d43fa537: "instance variable  declared with incompatible types in different translation units (2 vs. 1)"
0x7f48d43fa594: "non-type template parameter declared with incompatible types in different translation units ( vs. 2)"
0x7f48d43fa5f9: " method 2 has a different number of parameters in different translation units (1 vs. '_Tp')"
0x7f48d43fa655: " method 2 has a parameter with a different types in different translation units (1 vs. '_Tp')"


(gdb) x /20xs (0x7f48d43fa010 + 4511505832 - 200)
0x7f49e127bcf0: "ncurrent"
0x7f49e127bcf9: "expected  in OpenMP clause '2'"
0x7f49e127bd18: "modifier '' cannot be used along with modifier '2'"
0x7f49e127bd4b: "mapping of union members is not allowed"
0x7f49e127bd73: "incorrect reduction identifier, expected one of '+', '-', '*', '&', '"
0x7f49e127bdb9: ""
0x7f49e127bdba: ""
0x7f49e127bdbb: ""
0x7f49e127bdbc: ""
0x7f49e127bdbd: ""
0x7f49e127bdbe: ""
```

I was able to construct a testcase that triggers the same assertion failure (and in assertions-disabled builds, triggers the same unbounded memory consumption) with `clang` (no `clangd` involved).

header.h:
```c++
#include "bad_include.h"

template <bool, class = void> struct enable_if {};
template <class T> struct enable_if<true, T> { typedef T type; };
template <bool B, class T = void> using enable_if_t = typename enable_if<B, T>::type;

template <typename> struct meta { static constexpr int value = 0; };
template <> struct meta<int> { static constexpr int value = 1; };
template <> struct meta<float> { static constexpr int value = 2; };

namespace N {
inline namespace inner {

template <class T>
constexpr enable_if_t<meta<T>::value == 0, void> midpoint(T) {}

template <class U>
constexpr enable_if_t<meta<U>::value == 1, void> midpoint(U) {}

template <class F>
constexpr enable_if_t<meta<F>::value == 2, void> midpoint(F) {}

} // namespace inner
} // namespace N
```

source.cpp:
```c++
int x = N::something;
```

Commands to trigger:
```console
clang++ -Xclang -fallow-pch-with-compiler-errors -c -x c++-header -std=c++20 header.h -o header.h.pch
clang++ -Xclang -fallow-pch-with-compiler-errors -include-pch header.h.pch -c -std=c++20 source.cpp
```

Output:

```
header.h:1:10: fatal error: 'bad_include.h' file not found
    1 | #include "bad_include.h"
      |          ^~~~~~~~~~~~~~~
clang++: clang/include/clang/Basic/Diagnostic.h:1550: DiagnosticBuilder clang::DiagnosticsEngine::Report(SourceLocation, unsigned int): Assertion `CurDiagID == std::numeric_limits<unsigned>::max() && "Multiple diagnostics in flight at once!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: clang++ -Xclang -fallow-pch-with-compiler-errors -include-pch header.h.pch -c -std=c++20 source.cpp
1.      source.cpp:1:21: current parser token ';'
 #0 0x00007f4514ff0a3d llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) llvm/lib/Support/Unix/Signals.inc:723:11
 #1 0x00007f4514ff0f2b PrintStackTraceSignalHandler(void*) llvm/lib/Support/Unix/Signals.inc:798:1
 #2 0x00007f4514feefb6 llvm::sys::RunSignalHandlers() llvm/lib/Support/Signals.cpp:105:5
 #3 0x00007f4514ff02ce llvm::sys::CleanupOnSignal(unsigned long) llvm/lib/Support/Unix/Signals.inc:368:1
 #4 0x00007f4514ea2d07 (anonymous namespace)::CrashRecoveryContextImpl::HandleCrash(int, unsigned long) llvm/lib/Support/CrashRecoveryContext.cpp:0:7
 #5 0x00007f4514ea30a5 CrashRecoverySignalHandler(int) llvm/lib/Support/CrashRecoveryContext.cpp:391:1
 #6 0x00007f451da39140 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x13140)
 #7 0x00007f4514737ce1 raise ./signal/../sysdeps/unix/sysv/linux/raise.c:51:1
 #8 0x00007f4514721537 abort ./stdlib/abort.c:81:7
 #9 0x00007f451472140f get_sysdep_segment_value ./intl/loadmsgcat.c:509:8
#10 0x00007f451472140f _nl_load_domain ./intl/loadmsgcat.c:970:34
#11 0x00007f4514730662 (/lib/x86_64-linux-gnu/libc.so.6+0x31662)
#12 0x00007f450ff785d7 clang::DiagnosticsEngine::Report(clang::SourceLocation, unsigned int) clang/include/clang/Basic/Diagnostic.h:0:3
#13 0x00007f4510825875 clang::ODRDiagsEmitter::Diag(clang::SourceLocation, unsigned int) const clang/include/clang/AST/ODRDiagsEmitter.h:133:18
#14 0x00007f4510821adc clang::ODRDiagsEmitter::diagnoseMismatch(clang::FunctionDecl const*, clang::FunctionDecl const*) const::$_0::operator()(clang::SourceLocation, clang::SourceRange, clang::ODRDiagsEmitter::diagnoseMismatch(clang::FunctionDecl const*, clang::FunctionDecl const*) const::ODRFunctionDifference) const clang/lib/AST/ODRDiagsEmitter.cpp:1683:12
#15 0x00007f4510821329 clang::ODRDiagsEmitter::diagnoseMismatch(clang::FunctionDecl const*, clang::FunctionDecl const*) const clang/lib/AST/ODRDiagsEmitter.cpp:1738:17
#16 0x00007f4518f1edcf clang::ASTReader::diagnoseOdrViolations() clang/lib/Serialization/ASTReader.cpp:10100:11
#17 0x00007f4518f1fbbc clang::ASTReader::FinishedDeserializing() clang/lib/Serialization/ASTReader.cpp:10222:5
#18 0x00007f4518f32f2c clang::ExternalASTSource::Deserializing::~Deserializing() clang/include/clang/AST/ExternalASTSource.h:87:5
#19 0x00007f4518f16bd7 clang::ASTReader::FindExternalVisibleDeclsByName(clang::DeclContext const*, clang::DeclarationName) clang/lib/Serialization/ASTReader.cpp:8030:1
#20 0x00007f45103869a9 clang::DeclContext::lookup(clang::DeclarationName) const clang/lib/AST/DeclBase.cpp:1871:60
#21 0x00007f4510360657 clang::NamespaceDecl::isRedundantInlineQualifierFor(clang::DeclarationName) const clang/include/clang/AST/Decl.h:644:14
#22 0x00007f451034d70e clang::NamedDecl::printNestedNameSpecifier(llvm::raw_ostream&, clang::PrintingPolicy const&) const clang/lib/AST/Decl.cpp:1747:9
#23 0x00007f451034d34a clang::NamedDecl::printQualifiedName(llvm::raw_ostream&, clang::PrintingPolicy const&) const clang/lib/AST/Decl.cpp:1694:3
#24 0x00007f451034ddde clang::NamedDecl::getNameForDiagnostic(llvm::raw_ostream&, clang::PrintingPolicy const&, bool) const clang/lib/AST/Decl.cpp:1823:5
#25 0x00007f4510353508 clang::FunctionDecl::getNameForDiagnostic(llvm::raw_ostream&, clang::PrintingPolicy const&, bool) const clang/lib/AST/Decl.cpp:3079:14
#26 0x00007f450ffd149d clang::FormatASTNodeDiagnosticArgument(clang::DiagnosticsEngine::ArgumentKind, long, llvm::StringRef, llvm::StringRef, llvm::ArrayRef<std::pair<clang::DiagnosticsEngine::ArgumentKind, long>>, llvm::SmallVectorImpl<char>&, void*, llvm::ArrayRef<long>) clang/lib/AST/ASTDiagnostic.cpp:462:7
#27 0x00007f451889bea1 clang::DiagnosticsEngine::ConvertArgToString(clang::DiagnosticsEngine::ArgumentKind, long, llvm::StringRef, llvm::StringRef, llvm::ArrayRef<std::pair<clang::DiagnosticsEngine::ArgumentKind, long>>, llvm::SmallVectorImpl<char>&, llvm::ArrayRef<long>) const clang/include/clang/Basic/Diagnostic.h:887:3
#28 0x00007f45188975a8 clang::Diagnostic::FormatDiagnostic(char const*, char const*, llvm::SmallVectorImpl<char>&) const clang/lib/Basic/Diagnostic.cpp:1066:7
#29 0x00007f45188964ec clang::Diagnostic::FormatDiagnostic(llvm::SmallVectorImpl<char>&) const clang/lib/Basic/Diagnostic.cpp:802:1
#30 0x00007f4519cea18a clang::TextDiagnosticPrinter::HandleDiagnostic(clang::DiagnosticsEngine::Level, clang::Diagnostic const&) clang/lib/Frontend/TextDiagnosticPrinter.cpp:119:29
#31 0x00007f45188aa3d2 clang::DiagnosticIDs::EmitDiag(clang::DiagnosticsEngine&, clang::DiagnosticIDs::Level) const clang/lib/Basic/DiagnosticIDs.cpp:837:7
#32 0x00007f45188aa1f4 clang::DiagnosticIDs::ProcessDiag(clang::DiagnosticsEngine&) const clang/lib/Basic/DiagnosticIDs.cpp:829:3
#33 0x00007f451889b359 clang::DiagnosticsEngine::ProcessDiag() clang/include/clang/Basic/Diagnostic.h:1042:5
#34 0x00007f4518896319 clang::DiagnosticsEngine::EmitCurrentDiagnostic(bool) clang/lib/Basic/Diagnostic.cpp:546:13
#35 0x00007f450ff7902f clang::DiagnosticBuilder::Emit() clang/include/clang/Basic/Diagnostic.h:1325:28
#36 0x00007f450ff78519 clang::DiagnosticBuilder::~DiagnosticBuilder() clang/include/clang/Basic/Diagnostic.h:1366:26
#37 0x00007f451082137c clang::ODRDiagsEmitter::diagnoseMismatch(clang::FunctionDecl const*, clang::FunctionDecl const*) const clang/lib/AST/ODRDiagsEmitter.cpp:1742:15
#38 0x00007f4518f1edcf clang::ASTReader::diagnoseOdrViolations() clang/lib/Serialization/ASTReader.cpp:10100:11
#39 0x00007f4518f1fbbc clang::ASTReader::FinishedDeserializing() clang/lib/Serialization/ASTReader.cpp:10222:5
#40 0x00007f4518f32f2c clang::ExternalASTSource::Deserializing::~Deserializing() clang/include/clang/AST/ExternalASTSource.h:87:5
#41 0x00007f451905c1d2 clang::ASTReader::ReadDeclRecord(clang::GlobalDeclID) clang/lib/Serialization/ASTReaderDecl.cpp:4175:1
#42 0x00007f4518f0e46b clang::ASTReader::GetDecl(clang::GlobalDeclID) clang/lib/Serialization/ASTReader.cpp:7816:5
#43 0x00007f4518f16d98 clang::ASTReader::completeVisibleDeclsMap(clang::DeclContext const*) clang/lib/Serialization/ASTReader.cpp:8043:37
#44 0x00007f450dc4b5c8 clang::DeclContext::lookups() const clang/include/clang/AST/DeclLookups.h:79:29
#45 0x00007f450dc3a791 (anonymous namespace)::LookupVisibleHelper::lookupInDeclContext(clang::DeclContext*, clang::LookupResult&, bool, bool) clang/lib/Sema/SemaLookup.cpp:4225:10
#46 0x00007f450dc2a08f (anonymous namespace)::LookupVisibleHelper::lookupVisibleDecls(clang::Sema&, clang::DeclContext*, clang::Sema::LookupNameKind, bool) clang/lib/Sema/SemaLookup.cpp:4169:3
#47 0x00007f450dc29f32 clang::Sema::LookupVisibleDecls(clang::DeclContext*, clang::Sema::LookupNameKind, clang::VisibleDeclConsumer&, bool, bool, bool) clang/lib/Sema/SemaLookup.cpp:4475:1
#48 0x00007f450dc2d31a clang::Sema::makeTypoCorrectionConsumer(clang::DeclarationNameInfo const&, clang::Sema::LookupNameKind, clang::Scope*, clang::CXXScopeSpec*, clang::CorrectionCandidateCallback&, clang::DeclContext*, bool, clang::ObjCObjectPointerType const*, bool) clang/lib/Sema/SemaLookup.cpp:5323:3
#49 0x00007f450dc2f0fe clang::Sema::CorrectTypoDelayed(clang::DeclarationNameInfo const&, clang::Sema::LookupNameKind, clang::Scope*, clang::CXXScopeSpec*, clang::CorrectionCandidateCallback&, std::function<void (clang::TypoCorrection const&)>, std::function<clang::ActionResult<clang::Expr*, true> (clang::Sema&, clang::TypoExpr*, clang::TypoCorrection)>, clang::Sema::CorrectTypoKind, clang::DeclContext*, bool, clang::ObjCObjectPointerType const*) clang/lib/Sema/SemaLookup.cpp:5554:18
#50 0x00007f450d60eb52 clang::Sema::DiagnoseEmptyLookup(clang::Scope*, clang::CXXScopeSpec&, clang::LookupResult&, clang::CorrectionCandidateCallback&, clang::TemplateArgumentListInfo*, llvm::ArrayRef<clang::Expr*>, clang::DeclContext*, clang::TypoExpr**) clang/lib/Sema/SemaExpr.cpp:2537:12
#51 0x00007f450d606741 clang::Sema::ActOnIdExpression(clang::Scope*, clang::CXXScopeSpec&, clang::SourceLocation, clang::UnqualifiedId&, bool, bool, clang::CorrectionCandidateCallback*, bool, clang::Token*) clang/lib/Sema/SemaExpr.cpp:2826:9
#52 0x00007f450e7a190b clang::Parser::tryParseCXXIdExpression(clang::CXXScopeSpec&, bool, clang::Token&) clang/lib/Parse/ParseExprCXX.cpp:658:17
#53 0x00007f450e7a330a clang::Parser::ParseCXXIdExpression(bool) clang/lib/Parse/ParseExprCXX.cpp:751:7
#54 0x00007f450e78c94e clang::Parser::ParseCastExpression(clang::Parser::CastParseKind, bool, bool&, clang::Parser::TypeCastState, bool, bool*) clang/lib/Parse/ParseExpr.cpp:1683:11
#55 0x00007f450e78a756 clang::Parser::ParseCastExpression(clang::Parser::CastParseKind, bool, bool&, clang::Parser::TypeCastState, bool, bool*) clang/lib/Parse/ParseExpr.cpp:1198:18
#56 0x00007f450e784429 clang::Parser::ParseCastExpression(clang::Parser::CastParseKind, bool, clang::Parser::TypeCastState, bool, bool*) clang/lib/Parse/ParseExpr.cpp:712:20
#57 0x00007f450e782921 clang::Parser::ParseAssignmentExpression(clang::Parser::TypeCastState) clang/lib/Parse/ParseExpr.cpp:182:20
#58 0x00007f450e73524a clang::Parser::ParseInitializer() clang/include/clang/Parse/Parser.h:2115:14
#59 0x00007f450e71a163 clang::Parser::ParseDeclarationAfterDeclaratorAndAttributes(clang::Declarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::ForRangeInit*) clang/lib/Parse/ParseDecl.cpp:2764:25
#60 0x00007f450e7186f9 clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, clang::DeclaratorContext, clang::ParsedAttributes&, clang::Parser::ParsedTemplateInfo&, clang::SourceLocation*, clang::Parser::ForRangeInit*) clang/lib/Parse/ParseDecl.cpp:2479:9
#61 0x00007f450e85bb3c clang::Parser::ParseDeclOrFunctionDefInternal(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec&, clang::AccessSpecifier) clang/lib/Parse/Parser.cpp:1249:10
#62 0x00007f450e85b01e clang::Parser::ParseDeclarationOrFunctionDefinition(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*, clang::AccessSpecifier) clang/lib/Parse/Parser.cpp:1271:12
#63 0x00007f450e85a89a clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*) clang/lib/Parse/Parser.cpp:1074:14
#64 0x00007f450e8586dc clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) clang/lib/Parse/Parser.cpp:763:12
#65 0x00007f450e857d80 clang::Parser::ParseFirstTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) clang/lib/Parse/Parser.cpp:608:8
#66 0x00007f450e6f212b clang::ParseAST(clang::Sema&, bool, bool) clang/lib/Parse/ParseAST.cpp:162:15
#67 0x00007f4519c5d881 clang::ASTFrontendAction::ExecuteAction() clang/lib/Frontend/FrontendAction.cpp:1195:1
#68 0x00007f451d62e7e6 clang::CodeGenAction::ExecuteAction() clang/lib/CodeGen/CodeGenAction.cpp:1147:5
#69 0x00007f4519c5d27c clang::FrontendAction::Execute() clang/lib/Frontend/FrontendAction.cpp:1082:7
#70 0x00007f4519b53b89 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) clang/lib/Frontend/CompilerInstance.cpp:1061:23
#71 0x00007f451da0c73e clang::ExecuteCompilerInvocation(clang::CompilerInstance*) clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:280:8
#72 0x000055a18ab0d2a1 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) clang/tools/driver/cc1_main.cpp:232:13
#73 0x000055a18aaffc02 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) clang/tools/driver/driver.cpp:215:5
#74 0x000055a18ab006ad clang_main(int, char**, llvm::ToolContext const&)::$_0::operator()(llvm::SmallVectorImpl<char const*>&) const clang/tools/driver/driver.cpp:355:7
#75 0x000055a18ab0067d int llvm::function_ref<int (llvm::SmallVectorImpl<char const*>&)>::callback_fn<clang_main(int, char**, llvm::ToolContext const&)::$_0>(long, llvm::SmallVectorImpl<char const*>&) llvm/include/llvm/ADT/STLFunctionalExtras.h:45:5
#76 0x00007f45194c65b1 llvm::function_ref<int (llvm::SmallVectorImpl<char const*>&)>::operator()(llvm::SmallVectorImpl<char const*>&) const llvm/include/llvm/ADT/STLFunctionalExtras.h:68:5
#77 0x00007f45194c0638 clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::$_0::operator()() const clang/lib/Driver/Job.cpp:440:34
#78 0x00007f45194c0605 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::$_0>(long) llvm/include/llvm/ADT/STLFunctionalExtras.h:45:5
#79 0x00007f4514e8e829 llvm::function_ref<void ()>::operator()() const llvm/include/llvm/ADT/STLFunctionalExtras.h:68:5
#80 0x00007f4514ea2afa llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) llvm/lib/Support/CrashRecoveryContext.cpp:427:3
#81 0x00007f45194bff9b clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const clang/lib/Driver/Job.cpp:440:7
#82 0x00007f451944061f clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const clang/lib/Driver/Compilation.cpp:199:15
#83 0x00007f4519440827 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const clang/lib/Driver/Compilation.cpp:253:13
#84 0x00007f4519461f92 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) clang/lib/Driver/Driver.cpp:1950:7
#85 0x000055a18aaff6db clang_main(int, char**, llvm::ToolContext const&) clang/tools/driver/driver.cpp:391:9
#86 0x000055a18ab32b35 main tools/clang/tools/driver/clang-driver.cpp:17:3
#87 0x00007f4514722d0a __libc_start_main ./csu/../csu/libc-start.c:308:16
#88 0x000055a18aafe3fa _start (clang+++0x3c3fa)
clang++: error: clang frontend command failed with exit code 134 (use -v to see invocation)
clang version 19.0.0git
Target: x86_64-unknown-linux-gnu
Thread model: posix
Build config: +unoptimized, +assertions
clang++: note: diagnostic msg: Error generating preprocessed source(s).
```

--- a/clang-tools-extra/clang-tidy/ClangTidyDiagnosticConsumer.cpp
+++ b/clang-tools-extra/clang-tidy/ClangTidyDiagnosticConsumer.cpp
@@ -380,7 +380,6 @@ void ClangTidyDiagnosticConsumer::HandleDiagnostic(
     ++Context.Stats.ErrorsIgnoredNOLINT;
     // Ignored a warning, should ignore related notes as well
     LastErrorWasIgnored = true;
-    Context.DiagEngine->Clear();
     for (const auto &Error : SuppressionErrors)
       Context.diag(Error);
     return;
@@ -457,7 +456,6 @@ void ClangTidyDiagnosticConsumer::HandleDiagnostic(
   if (Info.hasSourceManager())
     checkFilters(Info.getLocation(), Info.getSourceManager());
 
-  Context.DiagEngine->Clear();
   for (const auto &Error : SuppressionErrors)
     Context.diag(Error);
 }

--- a/clang/include/clang/Basic/Diagnostic.h
+++ b/clang/include/clang/Basic/Diagnostic.h
@@ -964,8 +964,7 @@ private:
   // in DiagnosticBuilder in order to keep DiagnosticBuilder a small lightweight
   // object.  This implementation choice means that we can only have a few
   // diagnostics "in flight" at a time, but this seems to be a reasonable
-  // tradeoff to keep these objects small.  Assertions verify that only one
-  // diagnostic is in flight at a time.
+  // tradeoff to keep these objects small.
   friend class Diagnostic;
   friend class DiagnosticBuilder;
   friend class DiagnosticErrorTrap;
@@ -1196,14 +1195,8 @@ class DiagnosticBuilder : public StreamingDiagnostic {
 
   mutable DiagnosticsEngine *DiagObj = nullptr;
 
-  /// The location of the current diagnostic that is in flight.
-  SourceLocation CurDiagLoc;
-
-  /// The ID of the current diagnostic that is in flight.
-  ///
-  /// This is set to std::numeric_limits<unsigned>::max() when there is no
-  /// diagnostic in flight.
-  unsigned CurDiagID;
+  SourceLocation DiagLoc;
+  unsigned DiagID;
 
   /// Optional flag value.
   ///
@@ -1225,8 +1218,8 @@ class DiagnosticBuilder : public StreamingDiagnostic {
 
   DiagnosticBuilder() = default;
 
-  DiagnosticBuilder(DiagnosticsEngine *DiagObj, SourceLocation CurDiagLoc,
-                    unsigned CurDiagID);
+  DiagnosticBuilder(DiagnosticsEngine *DiagObj, SourceLocation DiagLoc,
+                    unsigned DiagID);
 
 protected:
   /// Clear out the current diagnostic.
@@ -1485,25 +1478,25 @@ inline DiagnosticBuilder DiagnosticsEngine::Report(unsigned DiagID) {
 //===----------------------------------------------------------------------===//
 
 /// A little helper class (which is basically a smart pointer that forwards
-/// info from DiagnosticsEngine) that allows clients to enquire about the
-/// currently in-flight diagnostic.
+/// info from DiagnosticsEngine and DiagnosticStorage) that allows clients to
+/// enquire about the diagnostic.
 class Diagnostic {
   const DiagnosticsEngine *DiagObj;
-  SourceLocation CurDiagLoc;
-  unsigned CurDiagID;
+  SourceLocation DiagLoc;
+  unsigned DiagID;
   std::string FlagValue;
   const DiagnosticStorage &DiagStorage;
   std::optional<StringRef> StoredDiagMessage;
 
 public:
   Diagnostic(const DiagnosticsEngine *DO, const DiagnosticBuilder &DiagBuilder);
-  Diagnostic(const DiagnosticsEngine *DO, SourceLocation CurDiagLoc,
-             unsigned CurDiagID, const DiagnosticStorage &DiagStorage,
+  Diagnostic(const DiagnosticsEngine *DO, SourceLocation DiagLoc,
+             unsigned DiagID, const DiagnosticStorage &DiagStorage,
              StringRef StoredDiagMessage);
 
   const DiagnosticsEngine *getDiags() const { return DiagObj; }
-  unsigned getID() const { return CurDiagID; }
-  const SourceLocation &getLocation() const { return CurDiagLoc; }
+  unsigned getID() const { return DiagID; }
+  const SourceLocation &getLocation() const { return DiagLoc; }
   bool hasSourceManager() const { return DiagObj->hasSourceManager(); }
   SourceManager &getSourceManager() const { return DiagObj->getSourceManager();}
 

--- a/clang/include/clang/Basic/Diagnostic.h
+++ b/clang/include/clang/Basic/Diagnostic.h
@@ -1225,7 +1225,7 @@ class DiagnosticBuilder : public StreamingDiagnostic {
 
   DiagnosticBuilder() = default;
 
-  DiagnosticBuilder(DiagnosticsEngine *diagObj, SourceLocation CurDiagLoc,
+  DiagnosticBuilder(DiagnosticsEngine *DiagObj, SourceLocation CurDiagLoc,
                     unsigned CurDiagID);
 
 protected:
@@ -1499,7 +1499,7 @@ public:
   Diagnostic(const DiagnosticsEngine *DO, const DiagnosticBuilder &DiagBuilder);
   Diagnostic(const DiagnosticsEngine *DO, SourceLocation CurDiagLoc,
              unsigned CurDiagID, const DiagnosticStorage &DiagStorage,
-             StringRef storedDiagMessage);
+             StringRef StoredDiagMessage);
 
   const DiagnosticsEngine *getDiags() const { return DiagObj; }
   unsigned getID() const { return CurDiagID; }

--- a/clang/include/clang/Basic/Diagnostic.h
+++ b/clang/include/clang/Basic/Diagnostic.h
@@ -183,6 +183,41 @@ struct DiagnosticStorage {
   DiagnosticStorage() = default;
 };
 
+/// An allocator for DiagnosticStorage objects, which uses a small cache to
+/// objects, used to reduce malloc()/free() traffic for partial diagnostics.
+class DiagStorageAllocator {
+  static const unsigned NumCached = 16;
+  DiagnosticStorage Cached[NumCached];
+  DiagnosticStorage *FreeList[NumCached];
+  unsigned NumFreeListEntries;
+
+public:
+  DiagStorageAllocator();
+  ~DiagStorageAllocator();
+
+  /// Allocate new storage.
+  DiagnosticStorage *Allocate() {
+    if (NumFreeListEntries == 0)
+      return new DiagnosticStorage;
+
+    DiagnosticStorage *Result = FreeList[--NumFreeListEntries];
+    Result->NumDiagArgs = 0;
+    Result->DiagRanges.clear();
+    Result->FixItHints.clear();
+    return Result;
+  }
+
+  /// Free the given storage object.
+  void Deallocate(DiagnosticStorage *S) {
+    if (S >= Cached && S <= Cached + NumCached) {
+      FreeList[NumFreeListEntries++] = S;
+      return;
+    }
+
+    delete S;
+  }
+};
+
 /// Concrete class used by the front-end to report problems and issues.
 ///
 /// This massages the diagnostics (e.g. handling things like "report warnings
@@ -519,27 +554,6 @@ private:
 
   void *ArgToStringCookie = nullptr;
   ArgToStringFnTy ArgToStringFn;
-
-  /// ID of the "delayed" diagnostic, which is a (typically
-  /// fatal) diagnostic that had to be delayed because it was found
-  /// while emitting another diagnostic.
-  unsigned DelayedDiagID;
-
-  /// First string argument for the delayed diagnostic.
-  std::string DelayedDiagArg1;
-
-  /// Second string argument for the delayed diagnostic.
-  std::string DelayedDiagArg2;
-
-  /// Third string argument for the delayed diagnostic.
-  std::string DelayedDiagArg3;
-
-  /// Optional flag value.
-  ///
-  /// Some flags accept values, for instance: -Wframe-larger-than=<value> and
-  /// -Rpass=<value>. The content of this string is emitted after the flag name
-  /// and '='.
-  std::string FlagValue;
 
 public:
   explicit DiagnosticsEngine(IntrusiveRefCntPtr<DiagnosticIDs> Diags,
@@ -945,50 +959,11 @@ public:
 
   void Report(const StoredDiagnostic &storedDiag);
 
-  /// Determine whethere there is already a diagnostic in flight.
-  bool isDiagnosticInFlight() const {
-    return CurDiagID != std::numeric_limits<unsigned>::max();
-  }
-
-  /// Set the "delayed" diagnostic that will be emitted once
-  /// the current diagnostic completes.
-  ///
-  ///  If a diagnostic is already in-flight but the front end must
-  ///  report a problem (e.g., with an inconsistent file system
-  ///  state), this routine sets a "delayed" diagnostic that will be
-  ///  emitted after the current diagnostic completes. This should
-  ///  only be used for fatal errors detected at inconvenient
-  ///  times. If emitting a delayed diagnostic causes a second delayed
-  ///  diagnostic to be introduced, that second delayed diagnostic
-  ///  will be ignored.
-  ///
-  /// \param DiagID The ID of the diagnostic being delayed.
-  ///
-  /// \param Arg1 A string argument that will be provided to the
-  /// diagnostic. A copy of this string will be stored in the
-  /// DiagnosticsEngine object itself.
-  ///
-  /// \param Arg2 A string argument that will be provided to the
-  /// diagnostic. A copy of this string will be stored in the
-  /// DiagnosticsEngine object itself.
-  ///
-  /// \param Arg3 A string argument that will be provided to the
-  /// diagnostic. A copy of this string will be stored in the
-  /// DiagnosticsEngine object itself.
-  void SetDelayedDiagnostic(unsigned DiagID, StringRef Arg1 = "",
-                            StringRef Arg2 = "", StringRef Arg3 = "");
-
-  /// Clear out the current diagnostic.
-  void Clear() { CurDiagID = std::numeric_limits<unsigned>::max(); }
-
-  /// Return the value associated with this diagnostic flag.
-  StringRef getFlagValue() const { return FlagValue; }
-
 private:
   // This is private state used by DiagnosticBuilder.  We put it here instead of
   // in DiagnosticBuilder in order to keep DiagnosticBuilder a small lightweight
-  // object.  This implementation choice means that we can only have one
-  // diagnostic "in flight" at a time, but this seems to be a reasonable
+  // object.  This implementation choice means that we can only have a few
+  // diagnostics "in flight" at a time, but this seems to be a reasonable
   // tradeoff to keep these objects small.  Assertions verify that only one
   // diagnostic is in flight at a time.
   friend class Diagnostic;
@@ -996,18 +971,6 @@ private:
   friend class DiagnosticErrorTrap;
   friend class DiagnosticIDs;
   friend class PartialDiagnostic;
-
-  /// Report the delayed diagnostic.
-  void ReportDelayed();
-
-  /// The location of the current diagnostic that is in flight.
-  SourceLocation CurDiagLoc;
-
-  /// The ID of the current diagnostic that is in flight.
-  ///
-  /// This is set to std::numeric_limits<unsigned>::max() when there is no
-  /// diagnostic in flight.
-  unsigned CurDiagID;
 
   enum {
     /// The maximum number of arguments we can hold.
@@ -1018,7 +981,7 @@ private:
     MaxArguments = DiagnosticStorage::MaxArguments,
   };
 
-  DiagnosticStorage DiagStorage;
+  DiagStorageAllocator DiagAllocator;
 
   DiagnosticMapping makeUserMapping(diag::Severity Map, SourceLocation L) {
     bool isPragma = L.isValid();
@@ -1038,8 +1001,8 @@ private:
   /// Used to report a diagnostic that is finally fully formed.
   ///
   /// \returns true if the diagnostic was emitted, false if it was suppressed.
-  bool ProcessDiag() {
-    return Diags->ProcessDiag(*this);
+  bool ProcessDiag(const DiagnosticBuilder &DiagBuilder) {
+    return Diags->ProcessDiag(*this, DiagBuilder);
   }
 
   /// @name Diagnostic Emission
@@ -1054,14 +1017,10 @@ protected:
   // Sema::Diag() patterns.
   friend class Sema;
 
-  /// Emit the current diagnostic and clear the diagnostic state.
+  /// Emit the diagnostic
   ///
   /// \param Force Emit the diagnostic regardless of suppression settings.
-  bool EmitCurrentDiagnostic(bool Force = false);
-
-  unsigned getCurrentDiagID() const { return CurDiagID; }
-
-  SourceLocation getCurrentDiagLoc() const { return CurDiagLoc; }
+  bool EmitDiagnostic(const DiagnosticBuilder &DB, bool Force = false);
 
   /// @}
 };
@@ -1114,40 +1073,7 @@ public:
 ///
 class StreamingDiagnostic {
 public:
-  /// An allocator for DiagnosticStorage objects, which uses a small cache to
-  /// objects, used to reduce malloc()/free() traffic for partial diagnostics.
-  class DiagStorageAllocator {
-    static const unsigned NumCached = 16;
-    DiagnosticStorage Cached[NumCached];
-    DiagnosticStorage *FreeList[NumCached];
-    unsigned NumFreeListEntries;
-
-  public:
-    DiagStorageAllocator();
-    ~DiagStorageAllocator();
-
-    /// Allocate new storage.
-    DiagnosticStorage *Allocate() {
-      if (NumFreeListEntries == 0)
-        return new DiagnosticStorage;
-
-      DiagnosticStorage *Result = FreeList[--NumFreeListEntries];
-      Result->NumDiagArgs = 0;
-      Result->DiagRanges.clear();
-      Result->FixItHints.clear();
-      return Result;
-    }
-
-    /// Free the given storage object.
-    void Deallocate(DiagnosticStorage *S) {
-      if (S >= Cached && S <= Cached + NumCached) {
-        FreeList[NumFreeListEntries++] = S;
-        return;
-      }
-
-      delete S;
-    }
-  };
+  using DiagStorageAllocator = clang::DiagStorageAllocator;
 
 protected:
   mutable DiagnosticStorage *DiagStorage = nullptr;
@@ -1236,11 +1162,6 @@ public:
 protected:
   StreamingDiagnostic() = default;
 
-  /// Construct with an external storage not owned by itself. The allocator
-  /// is a null pointer in this case.
-  explicit StreamingDiagnostic(DiagnosticStorage *Storage)
-      : DiagStorage(Storage) {}
-
   /// Construct with a storage allocator which will manage the storage. The
   /// allocator is not a null pointer in this case.
   explicit StreamingDiagnostic(DiagStorageAllocator &Alloc)
@@ -1271,8 +1192,25 @@ protected:
 class DiagnosticBuilder : public StreamingDiagnostic {
   friend class DiagnosticsEngine;
   friend class PartialDiagnostic;
+  friend class Diagnostic;
 
   mutable DiagnosticsEngine *DiagObj = nullptr;
+
+  /// The location of the current diagnostic that is in flight.
+  SourceLocation CurDiagLoc;
+
+  /// The ID of the current diagnostic that is in flight.
+  ///
+  /// This is set to std::numeric_limits<unsigned>::max() when there is no
+  /// diagnostic in flight.
+  unsigned CurDiagID;
+
+  /// Optional flag value.
+  ///
+  /// Some flags accept values, for instance: -Wframe-larger-than=<value> and
+  /// -Rpass=<value>. The content of this string is emitted after the flag name
+  /// and '='.
+  mutable std::string FlagValue;
 
   /// Status variable indicating if this diagnostic is still active.
   ///
@@ -1287,16 +1225,8 @@ class DiagnosticBuilder : public StreamingDiagnostic {
 
   DiagnosticBuilder() = default;
 
-  explicit DiagnosticBuilder(DiagnosticsEngine *diagObj)
-      : StreamingDiagnostic(&diagObj->DiagStorage), DiagObj(diagObj),
-        IsActive(true) {
-    assert(diagObj && "DiagnosticBuilder requires a valid DiagnosticsEngine!");
-    assert(DiagStorage &&
-           "DiagnosticBuilder requires a valid DiagnosticStorage!");
-    DiagStorage->NumDiagArgs = 0;
-    DiagStorage->DiagRanges.clear();
-    DiagStorage->FixItHints.clear();
-  }
+  DiagnosticBuilder(DiagnosticsEngine *diagObj, SourceLocation CurDiagLoc,
+                    unsigned CurDiagID);
 
 protected:
   /// Clear out the current diagnostic.
@@ -1322,7 +1252,7 @@ protected:
     if (!isActive()) return false;
 
     // Process the diagnostic.
-    bool Result = DiagObj->EmitCurrentDiagnostic(IsForceEmit);
+    bool Result = DiagObj->EmitDiagnostic(*this, IsForceEmit);
 
     // This diagnostic is dead.
     Clear();
@@ -1333,13 +1263,7 @@ protected:
 public:
   /// Copy constructor.  When copied, this "takes" the diagnostic info from the
   /// input and neuters it.
-  DiagnosticBuilder(const DiagnosticBuilder &D) : StreamingDiagnostic() {
-    DiagObj = D.DiagObj;
-    DiagStorage = D.DiagStorage;
-    IsActive = D.IsActive;
-    IsForceEmit = D.IsForceEmit;
-    D.Clear();
-  }
+  DiagnosticBuilder(const DiagnosticBuilder &D);
 
   template <typename T> const DiagnosticBuilder &operator<<(const T &V) const {
     assert(isActive() && "Clients must not add to cleared diagnostic!");
@@ -1371,7 +1295,7 @@ public:
     return *this;
   }
 
-  void addFlagValue(StringRef V) const { DiagObj->FlagValue = std::string(V); }
+  void addFlagValue(StringRef V) const { FlagValue = std::string(V); }
 };
 
 struct AddFlagValue {
@@ -1546,12 +1470,7 @@ const StreamingDiagnostic &operator<<(const StreamingDiagnostic &DB,
 
 inline DiagnosticBuilder DiagnosticsEngine::Report(SourceLocation Loc,
                                                    unsigned DiagID) {
-  assert(CurDiagID == std::numeric_limits<unsigned>::max() &&
-         "Multiple diagnostics in flight at once!");
-  CurDiagLoc = Loc;
-  CurDiagID = DiagID;
-  FlagValue.clear();
-  return DiagnosticBuilder(this);
+  return DiagnosticBuilder(this, Loc, DiagID);
 }
 
 const StreamingDiagnostic &operator<<(const StreamingDiagnostic &DB,
@@ -1570,20 +1489,25 @@ inline DiagnosticBuilder DiagnosticsEngine::Report(unsigned DiagID) {
 /// currently in-flight diagnostic.
 class Diagnostic {
   const DiagnosticsEngine *DiagObj;
+  SourceLocation CurDiagLoc;
+  unsigned CurDiagID;
+  std::string FlagValue;
+  const DiagnosticStorage &DiagStorage;
   std::optional<StringRef> StoredDiagMessage;
 
 public:
-  explicit Diagnostic(const DiagnosticsEngine *DO) : DiagObj(DO) {}
-  Diagnostic(const DiagnosticsEngine *DO, StringRef storedDiagMessage)
-      : DiagObj(DO), StoredDiagMessage(storedDiagMessage) {}
+  Diagnostic(const DiagnosticsEngine *DO, const DiagnosticBuilder &DiagBuilder);
+  Diagnostic(const DiagnosticsEngine *DO, SourceLocation CurDiagLoc,
+             unsigned CurDiagID, const DiagnosticStorage &DiagStorage,
+             StringRef storedDiagMessage);
 
   const DiagnosticsEngine *getDiags() const { return DiagObj; }
-  unsigned getID() const { return DiagObj->CurDiagID; }
-  const SourceLocation &getLocation() const { return DiagObj->CurDiagLoc; }
+  unsigned getID() const { return CurDiagID; }
+  const SourceLocation &getLocation() const { return CurDiagLoc; }
   bool hasSourceManager() const { return DiagObj->hasSourceManager(); }
   SourceManager &getSourceManager() const { return DiagObj->getSourceManager();}
 
-  unsigned getNumArgs() const { return DiagObj->DiagStorage.NumDiagArgs; }
+  unsigned getNumArgs() const { return DiagStorage.NumDiagArgs; }
 
   /// Return the kind of the specified index.
   ///
@@ -1593,8 +1517,7 @@ public:
   /// \pre Idx < getNumArgs()
   DiagnosticsEngine::ArgumentKind getArgKind(unsigned Idx) const {
     assert(Idx < getNumArgs() && "Argument index out of range!");
-    return (DiagnosticsEngine::ArgumentKind)
-        DiagObj->DiagStorage.DiagArgumentsKind[Idx];
+    return (DiagnosticsEngine::ArgumentKind)DiagStorage.DiagArgumentsKind[Idx];
   }
 
   /// Return the provided argument string specified by \p Idx.
@@ -1602,7 +1525,7 @@ public:
   const std::string &getArgStdStr(unsigned Idx) const {
     assert(getArgKind(Idx) == DiagnosticsEngine::ak_std_string &&
            "invalid argument accessor!");
-    return DiagObj->DiagStorage.DiagArgumentsStr[Idx];
+    return DiagStorage.DiagArgumentsStr[Idx];
   }
 
   /// Return the specified C string argument.
@@ -1610,8 +1533,7 @@ public:
   const char *getArgCStr(unsigned Idx) const {
     assert(getArgKind(Idx) == DiagnosticsEngine::ak_c_string &&
            "invalid argument accessor!");
-    return reinterpret_cast<const char *>(
-        DiagObj->DiagStorage.DiagArgumentsVal[Idx]);
+    return reinterpret_cast<const char *>(DiagStorage.DiagArgumentsVal[Idx]);
   }
 
   /// Return the specified signed integer argument.
@@ -1619,7 +1541,7 @@ public:
   int64_t getArgSInt(unsigned Idx) const {
     assert(getArgKind(Idx) == DiagnosticsEngine::ak_sint &&
            "invalid argument accessor!");
-    return (int64_t)DiagObj->DiagStorage.DiagArgumentsVal[Idx];
+    return (int64_t)DiagStorage.DiagArgumentsVal[Idx];
   }
 
   /// Return the specified unsigned integer argument.
@@ -1627,7 +1549,7 @@ public:
   uint64_t getArgUInt(unsigned Idx) const {
     assert(getArgKind(Idx) == DiagnosticsEngine::ak_uint &&
            "invalid argument accessor!");
-    return DiagObj->DiagStorage.DiagArgumentsVal[Idx];
+    return DiagStorage.DiagArgumentsVal[Idx];
   }
 
   /// Return the specified IdentifierInfo argument.
@@ -1636,7 +1558,7 @@ public:
     assert(getArgKind(Idx) == DiagnosticsEngine::ak_identifierinfo &&
            "invalid argument accessor!");
     return reinterpret_cast<IdentifierInfo *>(
-        DiagObj->DiagStorage.DiagArgumentsVal[Idx]);
+        DiagStorage.DiagArgumentsVal[Idx]);
   }
 
   /// Return the specified non-string argument in an opaque form.
@@ -1644,37 +1566,32 @@ public:
   uint64_t getRawArg(unsigned Idx) const {
     assert(getArgKind(Idx) != DiagnosticsEngine::ak_std_string &&
            "invalid argument accessor!");
-    return DiagObj->DiagStorage.DiagArgumentsVal[Idx];
+    return DiagStorage.DiagArgumentsVal[Idx];
   }
 
   /// Return the number of source ranges associated with this diagnostic.
-  unsigned getNumRanges() const {
-    return DiagObj->DiagStorage.DiagRanges.size();
-  }
+  unsigned getNumRanges() const { return DiagStorage.DiagRanges.size(); }
 
   /// \pre Idx < getNumRanges()
   const CharSourceRange &getRange(unsigned Idx) const {
     assert(Idx < getNumRanges() && "Invalid diagnostic range index!");
-    return DiagObj->DiagStorage.DiagRanges[Idx];
+    return DiagStorage.DiagRanges[Idx];
   }
 
   /// Return an array reference for this diagnostic's ranges.
-  ArrayRef<CharSourceRange> getRanges() const {
-    return DiagObj->DiagStorage.DiagRanges;
-  }
+  ArrayRef<CharSourceRange> getRanges() const { return DiagStorage.DiagRanges; }
 
-  unsigned getNumFixItHints() const {
-    return DiagObj->DiagStorage.FixItHints.size();
-  }
+  unsigned getNumFixItHints() const { return DiagStorage.FixItHints.size(); }
 
   const FixItHint &getFixItHint(unsigned Idx) const {
     assert(Idx < getNumFixItHints() && "Invalid index!");
-    return DiagObj->DiagStorage.FixItHints[Idx];
+    return DiagStorage.FixItHints[Idx];
   }
 
-  ArrayRef<FixItHint> getFixItHints() const {
-    return DiagObj->DiagStorage.FixItHints;
-  }
+  ArrayRef<FixItHint> getFixItHints() const { return DiagStorage.FixItHints; }
+
+  /// Return the value associated with this diagnostic flag.
+  StringRef getFlagValue() const { return FlagValue; }
 
   /// Format this diagnostic into a string, substituting the
   /// formal arguments into the %0 slots.

--- a/clang/include/clang/Basic/DiagnosticIDs.h
+++ b/clang/include/clang/Basic/DiagnosticIDs.h
@@ -22,6 +22,7 @@
 
 namespace clang {
   class DiagnosticsEngine;
+  class DiagnosticBuilder;
   class SourceLocation;
 
   // Import the diagnostic enums themselves.
@@ -367,11 +368,13 @@ private:
   ///
   /// \returns \c true if the diagnostic was emitted, \c false if it was
   /// suppressed.
-  bool ProcessDiag(DiagnosticsEngine &Diag) const;
+  bool ProcessDiag(DiagnosticsEngine &Diag,
+                   const DiagnosticBuilder &DiagBuilder) const;
 
   /// Used to emit a diagnostic that is finally fully formed,
   /// ignoring suppression.
-  void EmitDiag(DiagnosticsEngine &Diag, Level DiagLevel) const;
+  void EmitDiag(DiagnosticsEngine &Diag, const DiagnosticBuilder &DiagBuilder,
+                Level DiagLevel) const;
 
   /// Whether the diagnostic may leave the AST in a state where some
   /// invariants can break.

--- a/clang/include/clang/Basic/PartialDiagnostic.h
+++ b/clang/include/clang/Basic/PartialDiagnostic.h
@@ -166,13 +166,10 @@ public:
 
   void EmitToString(DiagnosticsEngine &Diags,
                     SmallVectorImpl<char> &Buf) const {
-    // FIXME: It should be possible to render a diagnostic to a string without
-    //        messing with the state of the diagnostics engine.
     DiagnosticBuilder DB(Diags.Report(getDiagID()));
     Emit(DB);
-    Diagnostic(&Diags).FormatDiagnostic(Buf);
+    Diagnostic(&Diags, DB).FormatDiagnostic(Buf);
     DB.Clear();
-    Diags.Clear();
   }
 
   /// Clear out this partial diagnostic, giving it a new diagnostic ID

--- a/clang/include/clang/Sema/Sema.h
+++ b/clang/include/clang/Sema/Sema.h
@@ -626,10 +626,10 @@ public:
   const llvm::MapVector<FieldDecl *, DeleteLocs> &
   getMismatchingDeleteExpressions() const;
 
-  /// Cause the active diagnostic on the DiagosticsEngine to be
-  /// emitted. This is closely coupled to the SemaDiagnosticBuilder class and
+  /// Cause the built diagnostic to be emitted on the DiagosticsEngine.
+  /// This is closely coupled to the SemaDiagnosticBuilder class and
   /// should not be used elsewhere.
-  void EmitCurrentDiagnostic(unsigned DiagID);
+  void EmitDiagnostic(unsigned DiagID, const DiagnosticBuilder &DB);
 
   void addImplicitTypedef(StringRef Name, QualType T);
 

--- a/clang/lib/Basic/Diagnostic.cpp
+++ b/clang/lib/Basic/Diagnostic.cpp
@@ -523,17 +523,16 @@ bool DiagnosticsEngine::EmitDiagnostic(const DiagnosticBuilder &DB,
 }
 
 DiagnosticBuilder::DiagnosticBuilder(DiagnosticsEngine *DiagObj,
-                                     SourceLocation CurDiagLoc,
-                                     unsigned CurDiagID)
+                                     SourceLocation DiagLoc, unsigned DiagID)
     : StreamingDiagnostic(DiagObj->DiagAllocator), DiagObj(DiagObj),
-      CurDiagLoc(CurDiagLoc), CurDiagID(CurDiagID), IsActive(true) {
+      DiagLoc(DiagLoc), DiagID(DiagID), IsActive(true) {
   assert(DiagObj && "DiagnosticBuilder requires a valid DiagnosticsEngine!");
 }
 
 DiagnosticBuilder::DiagnosticBuilder(const DiagnosticBuilder &D)
     : StreamingDiagnostic() {
-  CurDiagLoc = D.CurDiagLoc;
-  CurDiagID = D.CurDiagID;
+  DiagLoc = D.DiagLoc;
+  DiagID = D.DiagID;
   FlagValue = D.FlagValue;
   DiagObj = D.DiagObj;
   DiagStorage = D.DiagStorage;
@@ -546,14 +545,14 @@ DiagnosticBuilder::DiagnosticBuilder(const DiagnosticBuilder &D)
 
 Diagnostic::Diagnostic(const DiagnosticsEngine *DO,
                        const DiagnosticBuilder &DiagBuilder)
-    : DiagObj(DO), CurDiagLoc(DiagBuilder.CurDiagLoc),
-      CurDiagID(DiagBuilder.CurDiagID), FlagValue(DiagBuilder.FlagValue),
+    : DiagObj(DO), DiagLoc(DiagBuilder.DiagLoc),
+      DiagID(DiagBuilder.DiagID), FlagValue(DiagBuilder.FlagValue),
       DiagStorage(*DiagBuilder.getStorage()) {}
 
-Diagnostic::Diagnostic(const DiagnosticsEngine *DO, SourceLocation CurDiagLoc,
-                       unsigned CurDiagID, const DiagnosticStorage &DiagStorage,
+Diagnostic::Diagnostic(const DiagnosticsEngine *DO, SourceLocation DiagLoc,
+                       unsigned DiagID, const DiagnosticStorage &DiagStorage,
                        StringRef StoredDiagMessage)
-    : DiagObj(DO), CurDiagLoc(CurDiagLoc), CurDiagID(CurDiagID),
+    : DiagObj(DO), DiagLoc(DiagLoc), DiagID(DiagID),
       DiagStorage(DiagStorage), StoredDiagMessage(StoredDiagMessage) {}
 
 DiagnosticConsumer::~DiagnosticConsumer() = default;

--- a/clang/lib/Basic/Diagnostic.cpp
+++ b/clang/lib/Basic/Diagnostic.cpp
@@ -522,12 +522,12 @@ bool DiagnosticsEngine::EmitDiagnostic(const DiagnosticBuilder &DB,
   return Emitted;
 }
 
-DiagnosticBuilder::DiagnosticBuilder(DiagnosticsEngine *diagObj,
+DiagnosticBuilder::DiagnosticBuilder(DiagnosticsEngine *DiagObj,
                                      SourceLocation CurDiagLoc,
                                      unsigned CurDiagID)
-    : StreamingDiagnostic(diagObj->DiagAllocator), DiagObj(diagObj),
+    : StreamingDiagnostic(DiagObj->DiagAllocator), DiagObj(DiagObj),
       CurDiagLoc(CurDiagLoc), CurDiagID(CurDiagID), IsActive(true) {
-  assert(diagObj && "DiagnosticBuilder requires a valid DiagnosticsEngine!");
+  assert(DiagObj && "DiagnosticBuilder requires a valid DiagnosticsEngine!");
 }
 
 DiagnosticBuilder::DiagnosticBuilder(const DiagnosticBuilder &D)
@@ -552,9 +552,9 @@ Diagnostic::Diagnostic(const DiagnosticsEngine *DO,
 
 Diagnostic::Diagnostic(const DiagnosticsEngine *DO, SourceLocation CurDiagLoc,
                        unsigned CurDiagID, const DiagnosticStorage &DiagStorage,
-                       StringRef storedDiagMessage)
+                       StringRef StoredDiagMessage)
     : DiagObj(DO), CurDiagLoc(CurDiagLoc), CurDiagID(CurDiagID),
-      DiagStorage(DiagStorage), StoredDiagMessage(storedDiagMessage) {}
+      DiagStorage(DiagStorage), StoredDiagMessage(StoredDiagMessage) {}
 
 DiagnosticConsumer::~DiagnosticConsumer() = default;
 

--- a/clang/lib/Basic/Diagnostic.cpp
+++ b/clang/lib/Basic/Diagnostic.cpp
@@ -126,9 +126,7 @@ void DiagnosticsEngine::Reset(bool soft /*=false*/) {
   TrapNumErrorsOccurred = 0;
   TrapNumUnrecoverableErrorsOccurred = 0;
 
-  CurDiagID = std::numeric_limits<unsigned>::max();
   LastDiagLevel = DiagnosticIDs::Ignored;
-  DelayedDiagID = 0;
 
   if (!soft) {
     // Clear state related to #pragma diagnostic.
@@ -141,23 +139,6 @@ void DiagnosticsEngine::Reset(bool soft /*=false*/) {
     DiagStates.emplace_back();
     DiagStatesByLoc.appendFirst(&DiagStates.back());
   }
-}
-
-void DiagnosticsEngine::SetDelayedDiagnostic(unsigned DiagID, StringRef Arg1,
-                                             StringRef Arg2, StringRef Arg3) {
-  if (DelayedDiagID)
-    return;
-
-  DelayedDiagID = DiagID;
-  DelayedDiagArg1 = Arg1.str();
-  DelayedDiagArg2 = Arg2.str();
-  DelayedDiagArg3 = Arg3.str();
-}
-
-void DiagnosticsEngine::ReportDelayed() {
-  unsigned ID = DelayedDiagID;
-  DelayedDiagID = 0;
-  Report(ID) << DelayedDiagArg1 << DelayedDiagArg2 << DelayedDiagArg3;
 }
 
 DiagnosticMapping &
@@ -497,39 +478,31 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
 }
 
 void DiagnosticsEngine::Report(const StoredDiagnostic &storedDiag) {
-  assert(CurDiagID == std::numeric_limits<unsigned>::max() &&
-         "Multiple diagnostics in flight at once!");
-
-  CurDiagLoc = storedDiag.getLocation();
-  CurDiagID = storedDiag.getID();
-  DiagStorage.NumDiagArgs = 0;
-
-  DiagStorage.DiagRanges.clear();
+  DiagnosticStorage DiagStorage;
   DiagStorage.DiagRanges.append(storedDiag.range_begin(),
                                 storedDiag.range_end());
 
-  DiagStorage.FixItHints.clear();
   DiagStorage.FixItHints.append(storedDiag.fixit_begin(),
                                 storedDiag.fixit_end());
 
   assert(Client && "DiagnosticConsumer not set!");
   Level DiagLevel = storedDiag.getLevel();
-  Diagnostic Info(this, storedDiag.getMessage());
+  Diagnostic Info(this, storedDiag.getLocation(), storedDiag.getID(),
+                  DiagStorage, storedDiag.getMessage());
   Client->HandleDiagnostic(DiagLevel, Info);
   if (Client->IncludeInDiagnosticCounts()) {
     if (DiagLevel == DiagnosticsEngine::Warning)
       ++NumWarnings;
   }
-
-  CurDiagID = std::numeric_limits<unsigned>::max();
 }
 
-bool DiagnosticsEngine::EmitCurrentDiagnostic(bool Force) {
+bool DiagnosticsEngine::EmitDiagnostic(const DiagnosticBuilder &DB,
+                                       bool Force) {
   assert(getClient() && "DiagnosticClient not set!");
 
   bool Emitted;
   if (Force) {
-    Diagnostic Info(this);
+    Diagnostic Info(this, DB);
 
     // Figure out the diagnostic level of this message.
     DiagnosticIDs::Level DiagLevel
@@ -538,23 +511,50 @@ bool DiagnosticsEngine::EmitCurrentDiagnostic(bool Force) {
     Emitted = (DiagLevel != DiagnosticIDs::Ignored);
     if (Emitted) {
       // Emit the diagnostic regardless of suppression level.
-      Diags->EmitDiag(*this, DiagLevel);
+      Diags->EmitDiag(*this, DB, DiagLevel);
     }
   } else {
     // Process the diagnostic, sending the accumulated information to the
     // DiagnosticConsumer.
-    Emitted = ProcessDiag();
+    Emitted = ProcessDiag(DB);
   }
-
-  // Clear out the current diagnostic object.
-  Clear();
-
-  // If there was a delayed diagnostic, emit it now.
-  if (!Force && DelayedDiagID)
-    ReportDelayed();
 
   return Emitted;
 }
+
+DiagnosticBuilder::DiagnosticBuilder(DiagnosticsEngine *diagObj,
+                                     SourceLocation CurDiagLoc,
+                                     unsigned CurDiagID)
+    : StreamingDiagnostic(diagObj->DiagAllocator), DiagObj(diagObj),
+      CurDiagLoc(CurDiagLoc), CurDiagID(CurDiagID), IsActive(true) {
+  assert(diagObj && "DiagnosticBuilder requires a valid DiagnosticsEngine!");
+}
+
+DiagnosticBuilder::DiagnosticBuilder(const DiagnosticBuilder &D)
+    : StreamingDiagnostic() {
+  CurDiagLoc = D.CurDiagLoc;
+  CurDiagID = D.CurDiagID;
+  FlagValue = D.FlagValue;
+  DiagObj = D.DiagObj;
+  DiagStorage = D.DiagStorage;
+  D.DiagStorage = nullptr;
+  Allocator = D.Allocator;
+  IsActive = D.IsActive;
+  IsForceEmit = D.IsForceEmit;
+  D.Clear();
+}
+
+Diagnostic::Diagnostic(const DiagnosticsEngine *DO,
+                       const DiagnosticBuilder &DiagBuilder)
+    : DiagObj(DO), CurDiagLoc(DiagBuilder.CurDiagLoc),
+      CurDiagID(DiagBuilder.CurDiagID), FlagValue(DiagBuilder.FlagValue),
+      DiagStorage(*DiagBuilder.getStorage()) {}
+
+Diagnostic::Diagnostic(const DiagnosticsEngine *DO, SourceLocation CurDiagLoc,
+                       unsigned CurDiagID, const DiagnosticStorage &DiagStorage,
+                       StringRef storedDiagMessage)
+    : DiagObj(DO), CurDiagLoc(CurDiagLoc), CurDiagID(CurDiagID),
+      DiagStorage(DiagStorage), StoredDiagMessage(storedDiagMessage) {}
 
 DiagnosticConsumer::~DiagnosticConsumer() = default;
 
@@ -1210,13 +1210,13 @@ bool ForwardingDiagnosticConsumer::IncludeInDiagnosticCounts() const {
   return Target.IncludeInDiagnosticCounts();
 }
 
-PartialDiagnostic::DiagStorageAllocator::DiagStorageAllocator() {
+DiagStorageAllocator::DiagStorageAllocator() {
   for (unsigned I = 0; I != NumCached; ++I)
     FreeList[I] = Cached + I;
   NumFreeListEntries = NumCached;
 }
 
-PartialDiagnostic::DiagStorageAllocator::~DiagStorageAllocator() {
+DiagStorageAllocator::~DiagStorageAllocator() {
   // Don't assert if we are in a CrashRecovery context, as this invariant may
   // be invalidated during a crash.
   assert((NumFreeListEntries == NumCached ||

--- a/clang/lib/Basic/DiagnosticIDs.cpp
+++ b/clang/lib/Basic/DiagnosticIDs.cpp
@@ -571,7 +571,8 @@ DiagnosticIDs::getDiagnosticSeverity(unsigned DiagID, SourceLocation Loc,
   }
 
   // If explicitly requested, map fatal errors to errors.
-  if (Result == diag::Severity::Fatal && Diag.FatalsAsError)
+  if (Result == diag::Severity::Fatal &&
+      DiagID != diag::fatal_too_many_errors && Diag.FatalsAsError)
     Result = diag::Severity::Error;
 
   // Custom diagnostics always are emitted in system headers.

--- a/clang/lib/Basic/DiagnosticIDs.cpp
+++ b/clang/lib/Basic/DiagnosticIDs.cpp
@@ -571,8 +571,7 @@ DiagnosticIDs::getDiagnosticSeverity(unsigned DiagID, SourceLocation Loc,
   }
 
   // If explicitly requested, map fatal errors to errors.
-  if (Result == diag::Severity::Fatal &&
-      Diag.CurDiagID != diag::fatal_too_many_errors && Diag.FatalsAsError)
+  if (Result == diag::Severity::Fatal && Diag.FatalsAsError)
     Result = diag::Severity::Error;
 
   // Custom diagnostics always are emitted in system headers.
@@ -748,8 +747,9 @@ StringRef DiagnosticIDs::getNearestOption(diag::Flavor Flavor,
 
 /// ProcessDiag - This is the method used to report a diagnostic that is
 /// finally fully formed.
-bool DiagnosticIDs::ProcessDiag(DiagnosticsEngine &Diag) const {
-  Diagnostic Info(&Diag);
+bool DiagnosticIDs::ProcessDiag(DiagnosticsEngine &Diag,
+                                const DiagnosticBuilder &DiagBuilder) const {
+  Diagnostic Info(&Diag, DiagBuilder);
 
   assert(Diag.getClient() && "DiagnosticClient not set!");
 
@@ -815,22 +815,24 @@ bool DiagnosticIDs::ProcessDiag(DiagnosticsEngine &Diag) const {
     // stop a flood of bogus errors.
     if (Diag.ErrorLimit && Diag.NumErrors > Diag.ErrorLimit &&
         DiagLevel == DiagnosticIDs::Error) {
-      Diag.SetDelayedDiagnostic(diag::fatal_too_many_errors);
+      Diag.Report(diag::fatal_too_many_errors);
       return false;
     }
   }
 
   // Make sure we set FatalErrorOccurred to ensure that the notes from the
   // diagnostic that caused `fatal_too_many_errors` won't be emitted.
-  if (Diag.CurDiagID == diag::fatal_too_many_errors)
+  if (Info.getID() == diag::fatal_too_many_errors)
     Diag.FatalErrorOccurred = true;
   // Finally, report it.
-  EmitDiag(Diag, DiagLevel);
+  EmitDiag(Diag, DiagBuilder, DiagLevel);
   return true;
 }
 
-void DiagnosticIDs::EmitDiag(DiagnosticsEngine &Diag, Level DiagLevel) const {
-  Diagnostic Info(&Diag);
+void DiagnosticIDs::EmitDiag(DiagnosticsEngine &Diag,
+                             const DiagnosticBuilder &DiagBuilder,
+                             Level DiagLevel) const {
+  Diagnostic Info(&Diag, DiagBuilder);
   assert(DiagLevel != DiagnosticIDs::Ignored && "Cannot emit ignored diagnostics!");
 
   Diag.Client->HandleDiagnostic((DiagnosticsEngine::Level)DiagLevel, Info);
@@ -838,8 +840,6 @@ void DiagnosticIDs::EmitDiag(DiagnosticsEngine &Diag, Level DiagLevel) const {
     if (DiagLevel == DiagnosticIDs::Warning)
       ++Diag.NumWarnings;
   }
-
-  Diag.CurDiagID = ~0U;
 }
 
 bool DiagnosticIDs::isUnrecoverable(unsigned DiagID) const {

--- a/clang/lib/Basic/SourceManager.cpp
+++ b/clang/lib/Basic/SourceManager.cpp
@@ -130,13 +130,8 @@ ContentCache::getBufferOrNone(DiagnosticsEngine &Diag, FileManager &FM,
   // the file could also have been removed during processing. Since we can't
   // really deal with this situation, just create an empty buffer.
   if (!BufferOrError) {
-    if (Diag.isDiagnosticInFlight())
-      Diag.SetDelayedDiagnostic(diag::err_cannot_open_file,
-                                ContentsEntry->getName(),
-                                BufferOrError.getError().message());
-    else
-      Diag.Report(Loc, diag::err_cannot_open_file)
-          << ContentsEntry->getName() << BufferOrError.getError().message();
+    Diag.Report(Loc, diag::err_cannot_open_file)
+        << ContentsEntry->getName() << BufferOrError.getError().message();
 
     return std::nullopt;
   }
@@ -153,12 +148,7 @@ ContentCache::getBufferOrNone(DiagnosticsEngine &Diag, FileManager &FM,
   // ContentsEntry::getSize() could have the wrong size. Use
   // MemoryBuffer::getBufferSize() instead.
   if (Buffer->getBufferSize() >= std::numeric_limits<unsigned>::max()) {
-    if (Diag.isDiagnosticInFlight())
-      Diag.SetDelayedDiagnostic(diag::err_file_too_large,
-                                ContentsEntry->getName());
-    else
-      Diag.Report(Loc, diag::err_file_too_large)
-        << ContentsEntry->getName();
+    Diag.Report(Loc, diag::err_file_too_large) << ContentsEntry->getName();
 
     return std::nullopt;
   }
@@ -168,12 +158,7 @@ ContentCache::getBufferOrNone(DiagnosticsEngine &Diag, FileManager &FM,
   // have come from a stat cache).
   if (!ContentsEntry->isNamedPipe() &&
       Buffer->getBufferSize() != (size_t)ContentsEntry->getSize()) {
-    if (Diag.isDiagnosticInFlight())
-      Diag.SetDelayedDiagnostic(diag::err_file_modified,
-                                ContentsEntry->getName());
-    else
-      Diag.Report(Loc, diag::err_file_modified)
-        << ContentsEntry->getName();
+    Diag.Report(Loc, diag::err_file_modified) << ContentsEntry->getName();
 
     return std::nullopt;
   }

--- a/clang/lib/Frontend/Rewrite/FixItRewriter.cpp
+++ b/clang/lib/Frontend/Rewrite/FixItRewriter.cpp
@@ -200,10 +200,8 @@ void FixItRewriter::HandleDiagnostic(DiagnosticsEngine::Level DiagLevel,
 /// Emit a diagnostic via the adapted diagnostic client.
 void FixItRewriter::Diag(SourceLocation Loc, unsigned DiagID) {
   // When producing this diagnostic, we temporarily bypass ourselves,
-  // clear out any current diagnostic, and let the downstream client
-  // format the diagnostic.
+  // and let the downstream client format the diagnostic.
   Diags.setClient(Client, false);
-  Diags.Clear();
   Diags.Report(Loc, DiagID);
   Diags.setClient(this, false);
 }

--- a/clang/lib/Frontend/TextDiagnosticPrinter.cpp
+++ b/clang/lib/Frontend/TextDiagnosticPrinter.cpp
@@ -80,7 +80,7 @@ static void printDiagnosticOptions(raw_ostream &OS,
     if (!Opt.empty()) {
       OS << (Started ? "," : " [")
          << (Level == DiagnosticsEngine::Remark ? "-R" : "-W") << Opt;
-      StringRef OptValue = Info.getDiags()->getFlagValue();
+      StringRef OptValue = Info.getFlagValue();
       if (!OptValue.empty())
         OS << "=" << OptValue;
       Started = true;

--- a/clang/lib/Sema/Sema.cpp
+++ b/clang/lib/Sema/Sema.cpp
@@ -1589,7 +1589,7 @@ LangAS Sema::getDefaultCXXMethodAddrSpace() const {
   return LangAS::Default;
 }
 
-void Sema::EmitCurrentDiagnostic(unsigned DiagID) {
+void Sema::EmitDiagnostic(unsigned DiagID, const DiagnosticBuilder &DB) {
   // FIXME: It doesn't make sense to me that DiagID is an incoming argument here
   // and yet we also use the current diag ID on the DiagnosticsEngine. This has
   // been made more painfully obvious by the refactor that introduced this
@@ -1597,9 +1597,9 @@ void Sema::EmitCurrentDiagnostic(unsigned DiagID) {
   // eliminated. If it truly cannot be (for example, there is some reentrancy
   // issue I am not seeing yet), then there should at least be a clarifying
   // comment somewhere.
+  Diagnostic DiagInfo(&Diags, DB);
   if (std::optional<TemplateDeductionInfo *> Info = isSFINAEContext()) {
-    switch (DiagnosticIDs::getDiagnosticSFINAEResponse(
-              Diags.getCurrentDiagID())) {
+    switch (DiagnosticIDs::getDiagnosticSFINAEResponse(DiagInfo.getID())) {
     case DiagnosticIDs::SFINAE_Report:
       // We'll report the diagnostic below.
       break;
@@ -1612,13 +1612,11 @@ void Sema::EmitCurrentDiagnostic(unsigned DiagID) {
       // Make a copy of this suppressed diagnostic and store it with the
       // template-deduction information.
       if (*Info && !(*Info)->hasSFINAEDiagnostic()) {
-        Diagnostic DiagInfo(&Diags);
         (*Info)->addSFINAEDiagnostic(DiagInfo.getLocation(),
                        PartialDiagnostic(DiagInfo, Context.getDiagAllocator()));
       }
 
       Diags.setLastDiagnosticIgnored(true);
-      Diags.Clear();
       return;
 
     case DiagnosticIDs::SFINAE_AccessControl: {
@@ -1629,7 +1627,7 @@ void Sema::EmitCurrentDiagnostic(unsigned DiagID) {
       if (!AccessCheckingSFINAE && !getLangOpts().CPlusPlus11)
         break;
 
-      SourceLocation Loc = Diags.getCurrentDiagLoc();
+      SourceLocation Loc = DiagInfo.getLocation();
 
       // Suppress this diagnostic.
       ++NumSFINAEErrors;
@@ -1637,16 +1635,13 @@ void Sema::EmitCurrentDiagnostic(unsigned DiagID) {
       // Make a copy of this suppressed diagnostic and store it with the
       // template-deduction information.
       if (*Info && !(*Info)->hasSFINAEDiagnostic()) {
-        Diagnostic DiagInfo(&Diags);
         (*Info)->addSFINAEDiagnostic(DiagInfo.getLocation(),
                        PartialDiagnostic(DiagInfo, Context.getDiagAllocator()));
       }
 
       Diags.setLastDiagnosticIgnored(true);
-      Diags.Clear();
 
-      // Now the diagnostic state is clear, produce a C++98 compatibility
-      // warning.
+      // Now produce a C++98 compatibility warning.
       Diag(Loc, diag::warn_cxx98_compat_sfinae_access_control);
 
       // The last diagnostic which Sema produced was ignored. Suppress any
@@ -1659,14 +1654,12 @@ void Sema::EmitCurrentDiagnostic(unsigned DiagID) {
       // Make a copy of this suppressed diagnostic and store it with the
       // template-deduction information;
       if (*Info) {
-        Diagnostic DiagInfo(&Diags);
         (*Info)->addSuppressedDiagnostic(DiagInfo.getLocation(),
                        PartialDiagnostic(DiagInfo, Context.getDiagAllocator()));
       }
 
       // Suppress this diagnostic.
       Diags.setLastDiagnosticIgnored(true);
-      Diags.Clear();
       return;
     }
   }
@@ -1676,7 +1669,7 @@ void Sema::EmitCurrentDiagnostic(unsigned DiagID) {
   Context.setPrintingPolicy(getPrintingPolicy());
 
   // Emit the diagnostic.
-  if (!Diags.EmitCurrentDiagnostic())
+  if (!Diags.EmitDiagnostic(DB))
     return;
 
   // If this is not a note, and we're in a template instantiation

--- a/clang/lib/Sema/SemaBase.cpp
+++ b/clang/lib/Sema/SemaBase.cpp
@@ -26,7 +26,7 @@ SemaBase::ImmediateDiagBuilder::~ImmediateDiagBuilder() {
   Clear();
 
   // Dispatch to Sema to emit the diagnostic.
-  SemaRef.EmitCurrentDiagnostic(DiagID);
+  SemaRef.EmitDiagnostic(DiagID, *this);
 }
 
 PartialDiagnostic SemaBase::PDiag(unsigned DiagID) {

--- a/clang/lib/Serialization/ASTReader.cpp
+++ b/clang/lib/Serialization/ASTReader.cpp
@@ -1382,7 +1382,7 @@ bool ASTReader::ReadVisibleDeclContextStorage(ModuleFile &M,
 
 void ASTReader::Error(StringRef Msg) const {
   Error(diag::err_fe_pch_malformed, Msg);
-  if (PP.getLangOpts().Modules && !Diags.isDiagnosticInFlight() &&
+  if (PP.getLangOpts().Modules &&
       !PP.getHeaderSearchInfo().getModuleCachePath().empty()) {
     Diag(diag::note_module_cache_path)
       << PP.getHeaderSearchInfo().getModuleCachePath();
@@ -1391,10 +1391,7 @@ void ASTReader::Error(StringRef Msg) const {
 
 void ASTReader::Error(unsigned DiagID, StringRef Arg1, StringRef Arg2,
                       StringRef Arg3) const {
-  if (Diags.isDiagnosticInFlight())
-    Diags.SetDelayedDiagnostic(DiagID, Arg1, Arg2, Arg3);
-  else
-    Diag(DiagID) << Arg1 << Arg2 << Arg3;
+  Diag(DiagID) << Arg1 << Arg2 << Arg3;
 }
 
 void ASTReader::Error(llvm::Error &&Err) const {
@@ -2713,7 +2710,7 @@ InputFile ASTReader::getInputFile(ModuleFile &F, unsigned ID, bool Complain) {
 
   // For an overridden file, there is nothing to validate.
   if (!Overridden && FileChange.Kind != Change::None) {
-    if (Complain && !Diags.isDiagnosticInFlight()) {
+    if (Complain) {
       // Build a list of the PCH imports that got us here (in reverse).
       SmallVector<ModuleFile *, 4> ImportStack(1, &F);
       while (!ImportStack.back()->ImportedBy.empty())
@@ -3689,10 +3686,8 @@ llvm::Error ASTReader::ReadASTBlock(ModuleFile &F,
           SourceMgr.AllocateLoadedSLocEntries(F.LocalNumSLocEntries,
                                               SLocSpaceSize);
       if (!F.SLocEntryBaseID) {
-        if (!Diags.isDiagnosticInFlight()) {
-          Diags.Report(SourceLocation(), diag::remark_sloc_usage);
-          SourceMgr.noteSLocAddressSpaceUsage(Diags);
-        }
+        Diags.Report(SourceLocation(), diag::remark_sloc_usage);
+        SourceMgr.noteSLocAddressSpaceUsage(Diags);
         return llvm::createStringError(std::errc::invalid_argument,
                                        "ran out of source locations");
       }

--- a/clang/test/PCH/race-condition.cpp
+++ b/clang/test/PCH/race-condition.cpp
@@ -1,0 +1,41 @@
+// RUN: %clang_cc1 -fallow-pch-with-compiler-errors -std=c++20 -x c++-header -emit-pch %s -o %t -verify
+// RUN: %clang_cc1 -fallow-pch-with-compiler-errors -std=c++20 -include-pch %t %s -verify
+#ifndef HEADER_H
+#define HEADER_H
+
+#include "bad_include.h"
+// expected-error@6{{'bad_include.h' file not found}}
+
+template <bool, class = void> struct enable_if {};
+template <class T> struct enable_if<true, T> { typedef T type; };
+template <bool B, class T = void> using enable_if_t = typename enable_if<B, T>::type;
+
+template <typename> struct meta { static constexpr int value = 0; };
+template <> struct meta<int> { static constexpr int value = 1; };
+template <> struct meta<float> { static constexpr int value = 2; };
+
+namespace N {
+inline namespace inner {
+
+template <class T>
+constexpr enable_if_t<meta<T>::value == 0, void> midpoint(T) {}
+
+template <class U>
+constexpr enable_if_t<meta<U>::value == 1, void> midpoint(U) {}
+
+template <class F>
+constexpr enable_if_t<meta<F>::value == 2, void> midpoint(F) {}
+
+} // namespace inner
+} // namespace N
+
+#else
+
+// expected-error@27{{'N::midpoint' has different definitions in different modules; defined here first difference is 1st parameter with type 'F'}}
+// expected-error@24{{'N::midpoint' has different definitions in different modules; defined here first difference is 1st parameter with type 'U'}}
+// expected-note@21{{but in '' found 1st parameter with type 'T'}}
+int x = N::something;
+// expected-error@37{{no member named 'something' in namespace 'N'}}
+// expected-note@21{{but in '' found 1st parameter with type 'T'}}
+
+#endif

--- a/clang/unittests/Basic/DiagnosticTest.cpp
+++ b/clang/unittests/Basic/DiagnosticTest.cpp
@@ -80,6 +80,21 @@ TEST(DiagnosticTest, fatalsAsError) {
   }
 }
 
+TEST(DiagnosticTest, tooManyErrorsIsAlwaysFatal) {
+  DiagnosticsEngine Diags(new DiagnosticIDs(), new DiagnosticOptions,
+                          new IgnoringDiagConsumer());
+  Diags.setFatalsAsError(true);
+
+  // Report a fatal_too_many_errors diagnostic to ensure that still
+  // acts as a fatal error despite downgrading fatal errors to errors.
+  Diags.Report(diag::fatal_too_many_errors);
+  EXPECT_TRUE(Diags.hasFatalErrorOccurred());
+
+  // Ensure that the severity of that diagnostic is really "fatal".
+  EXPECT_EQ(Diags.getDiagnosticLevel(diag::fatal_too_many_errors, {}),
+            DiagnosticsEngine::Level::Fatal);
+}
+
 // Check that soft RESET works as intended
 TEST(DiagnosticTest, softReset) {
   DiagnosticsEngine Diags(new DiagnosticIDs(), new DiagnosticOptions,

--- a/clang/unittests/Basic/DiagnosticTest.cpp
+++ b/clang/unittests/Basic/DiagnosticTest.cpp
@@ -17,9 +17,6 @@ using namespace llvm;
 using namespace clang;
 
 void clang::DiagnosticsTestHelper(DiagnosticsEngine &diag) {
-  unsigned delayedDiagID = 0U;
-
-  EXPECT_EQ(diag.DelayedDiagID, delayedDiagID);
   EXPECT_FALSE(diag.DiagStates.empty());
   EXPECT_TRUE(diag.DiagStatesByLoc.empty());
   EXPECT_TRUE(diag.DiagStateOnPushStack.empty());
@@ -104,7 +101,6 @@ TEST(DiagnosticTest, softReset) {
   // Check for private variables of DiagnosticsEngine differentiating soft reset
   DiagnosticsTestHelper(Diags);
 
-  EXPECT_FALSE(Diags.isDiagnosticInFlight());
   EXPECT_TRUE(Diags.isLastDiagnosticIgnored());
 }
 

--- a/clang/unittests/Driver/DXCModeTest.cpp
+++ b/clang/unittests/Driver/DXCModeTest.cpp
@@ -51,7 +51,6 @@ static void validateTargetProfile(
   EXPECT_TRUE(C);
   EXPECT_EQ(Diags.getNumErrors(), NumOfErrors);
   EXPECT_STREQ(DiagConsumer->Errors.back().c_str(), ExpectError.data());
-  Diags.Clear();
   DiagConsumer->clear();
 }
 
@@ -160,7 +159,6 @@ TEST(DxcModeTest, ValidatorVersionValidation) {
       DiagConsumer->Errors.back().c_str(),
       "invalid validator version : 0.1; if validator major version is 0, "
       "minor version must also be 0");
-  Diags.Clear();
   DiagConsumer->clear();
 
   Args = TheDriver.ParseArgStrings({"-validator-version", "1"}, false,
@@ -176,7 +174,6 @@ TEST(DxcModeTest, ValidatorVersionValidation) {
   EXPECT_STREQ(DiagConsumer->Errors.back().c_str(),
                "invalid validator version : 1; format of validator version is "
                "\"<major>.<minor>\" (ex:\"1.4\")");
-  Diags.Clear();
   DiagConsumer->clear();
 
   Args = TheDriver.ParseArgStrings({"-validator-version", "-Tlib_6_7"}, false,
@@ -193,7 +190,6 @@ TEST(DxcModeTest, ValidatorVersionValidation) {
       DiagConsumer->Errors.back().c_str(),
       "invalid validator version : -Tlib_6_7; format of validator version is "
       "\"<major>.<minor>\" (ex:\"1.4\")");
-  Diags.Clear();
   DiagConsumer->clear();
 
   Args = TheDriver.ParseArgStrings({"-validator-version", "foo"}, false,
@@ -210,7 +206,6 @@ TEST(DxcModeTest, ValidatorVersionValidation) {
       DiagConsumer->Errors.back().c_str(),
       "invalid validator version : foo; format of validator version is "
       "\"<major>.<minor>\" (ex:\"1.4\")");
-  Diags.Clear();
   DiagConsumer->clear();
 }
 

